readme.md
# PyWinCLI

PyWinCLI is a CLI for managing Python environments in a Wine prefix.

## Usage

`pywinc <option>`

Option           | Description
:--------------- | :----------
`-t`, `terminal` | Open PyWinCLI in Linux terminal
`-c`, `console`  | Open PyWinCLI in Wine console
`-w`, `wcfg`     | Run winecfg for PyWinCLI prefix
`-h`, `help`     | Display this help message

Once PyWinCLI is loaded, use the `pve` command to load the default Python virtual environment or run with one of the options listed below.

`pve <option>`

Option     | Description
:--------- | :----------
`-l`       | List available python environments
`-a <env>` | Activate selected environment
`-d`       | Deactivate loaded environment
`-r <env>` | Remove selected environment

## Installation

Run the `installer.sh` script to install PyWinCLI.  
Running `installer.sh` with the `-u` flag will uninstall PyWinCLI.

The installer will:

- Create a Wine prefix (default: `$HOME/.pywinc/pfx`).
- Create a Python environment directory (default: `$HOME/.pywinc/venv`).
- Install [Python 3.11.1](https://www.python.org/ftp/python/3.11.1/python-3.11.1-amd64.exe) to Wine prefix.
	+ **Required Options:**
      - Add python.exe to PATH
      - pip
      - tcl/tk and IDLe
- Install [NSIS 3.08](https://versaweb.dl.sourceforge.net/project/nsis/NSIS%203/3.08/nsis-3.08-setup.exe) to Wine prefix (*Optional: necessary for compiling with pynsist*).
  + **Required Options:**
      - User Interfaces
      - Graphics
      - Tools
      - Plugins
  + It is not recommended to run application after installation is complete.


changelog.md
# Changelog

## [v0.2] - Build 2022.1230.01

### Changed

- App name from PyWin to PyWinCLI
- Updated documentation

## [v0.1] - Build 2022.1230.00

### Added

- New functions:
	+ `deactivate`, `help` to `pve.bat`
	+ `help` to `pywin.sh`
	+ `uninstall` to `installer.sh`

### Changed

- Updated documentation

### Removed

- Configuration file support


pywin.sh
#! /bin/bash

# PyWin

## Manage python environments for wine.

## Global variables
PROFILE_DIR=$HOME/.pywin
SCRIPTS=$PROFILE_DIR/scripts
TMPDIR=$PROFILE_DIR/tmp
VENV="$PROFILE_DIR/venv"
PFX="$PROFILE_DIR/pfx"
CWD="$(pwd)"
CMD="$1"

## Wine settings
export WINEDEBUG=-all
export WINEDLLOVERRIDES=winemenubuilder.exe=d
export WINEPATH="$SCRIPTS"
export WINEPREFIX="$PFX"

## Functions

### Run Winecfg for prefix
wcfg(){
  winecfg
}

### Open Python prefix in bash terminal
terminal(){
  wine cmd
}

### Open Python prefix in wine console
console(){
  wineconsole
}

help(){
  echo -e "
  pywin <option>

  Option           | Description
  :--------------- | :----------
  -t, terminal | Open PyWin in Linux terminal
  -c, console  | Open PyWin in Wine console
  -w, wcfg     | Run winecfg for PyWin prefix
  -h, help     | Display this help message

  Once PyWin is loaded, use the pve command to load the default Python virtual environment or launch with one of the options listed below.

  pve <option>

  Option     | Description
  :--------- | :----------
  -l       | List available python environments
  -a <env> | Activate selected environment
  -d       | Deactivate loaded environment
  -r <env> | Remove selected environment
"
}

## Create environment if missing
if [[ ! -d $PFX ]]; then
  echo -e "\nPlease run PyWin install.sh script.\n"
fi

## Function shortcuts
if [[ -z $CMD ]]; then
  echo -e "\nRun pywin -h for help.\n"
elif [[ $CMD == "-h" ]]; then
  help
elif [[ $CMD == "-t" ]]; then
  terminal
elif [[ $CMD == "-c" ]]; then
  console
elif [[ $CMD == "-w" ]]; then
  wcfg
else
  $CMD
fi


pywinc.sh
#! /bin/bash

# PyWinCLI

## Manage python environments for wine.

## Global variables
PROFILE_DIR=$HOME/.pywinc
SCRIPTS=$PROFILE_DIR/scripts
TMPDIR=$PROFILE_DIR/tmp
VENV="$PROFILE_DIR/venv"
PFX="$PROFILE_DIR/pfx"
CWD="$(pwd)"
CMD="$1"

## Wine settings
export WINEDEBUG=-all
export WINEDLLOVERRIDES=winemenubuilder.exe=d
export WINEPATH="$SCRIPTS"
export WINEPREFIX="$PFX"

## Functions

### Run Winecfg for prefix
wcfg(){
  winecfg
}

### Open Python prefix in bash terminal
terminal(){
  wine cmd
}

### Open Python prefix in wine console
console(){
  wineconsole
}

help(){
  echo -e "
  pywinc <option>

  Option       | Description
  :----------- | :----------
  -t, terminal | Open PyWinCLI in Linux terminal
  -c, console  | Open PyWinCLI in Wine console
  -w, wcfg     | Run winecfg for PyWinCLI prefix
  -h, help     | Display this help message

  Once PyWinCLI is loaded, use the pve command to load the default Python virtual environment or launch with one of the options listed below.

  pve <option>

  Option   | Description
  :------- | :----------
  -l       | List available python environments
  -a <env> | Activate selected environment
  -d       | Deactivate loaded environment
  -r <env> | Remove selected environment
"
}

## Create environment if missing
if [[ ! -d $PFX ]]; then
  echo -e "\nPlease run PyWinCLI install.sh script.\n"
fi

## Function shortcuts
if [[ -z $CMD ]]; then
  echo -e "\nRun pywinc -h for help.\n"
elif [[ $CMD == "-h" ]]; then
  help
elif [[ $CMD == "-t" ]]; then
  terminal
elif [[ $CMD == "-c" ]]; then
  console
elif [[ $CMD == "-w" ]]; then
  wcfg
else
  $CMD
fi


install.sh
#! /bin/bash

# PyWin installer

## Global variables
SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
source "$SCRIPT_DIR/pywin.sh"
CMD=$1

## Functions

### Install PyWin
install()(
  #### Create profile directory
  mkdir -p "$TMPDIR"
  mkdir -p "$VENV"
  cd "$TMPDIR"

  #### Create Win 10/64-bit wine prefix
  winetricks -q mimeassoc=off win10

  #### Install Python 3.11.1 (64-Bit)
  wget "https://www.python.org/ftp/python/3.11.1/python-3.11.1-amd64.exe"
  if [[ -f python-3.11.1-amd64.exe ]]; then
    wine python-3.11.1-amd64.exe
  else
    echo "Please download python-3.11.1-amd64.exe to this directory."
  fi

  #### Install NSIS 3.08
  wget "https://versaweb.dl.sourceforge.net/project/nsis/NSIS%203/3.08/nsis-3.08-setup.exe"
  if [[ -f nsis-3.08-setup.exe ]]; then
    wine nsis-3.08-setup.exe
  else
    echo "Please download nsis-3.08-setup.exe to this directory."
  fi

  #### Install scripts
  mkdir -p "$SCRIPTS"
  cp -rv "$SCRIPT_DIR/pve.bat" "$SCRIPTS/"
  ln -s "$SCRIPT_DIR/pywin.sh" "$HOME/.local/bin/pywin"
  rm -rf "$TMPDIR"
)

### Uninstall PyWin
uninstall(){
  rm -rf "$PROFILE_DIR"  "$HOME/.local/bin/pywin" "$SCRIPT_DIR"
}

if [[ -z $CMD ]]; then
  install
else
  $CMD
fi


installer.sh
#! /bin/bash

# PyWinCLI installer

## Global variables
SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
source "$SCRIPT_DIR/pywinc.sh"
CMD=$1

## Functions

### Install PyWinCLI
install()(
  #### Create profile directory
  mkdir -p "$TMPDIR"
  mkdir -p "$VENV"
  cd "$TMPDIR"

  #### Create Win 10/64-bit wine prefix
  winetricks -q mimeassoc=off win10

  #### Install Python 3.11.1 (64-Bit)
  wget "https://www.python.org/ftp/python/3.11.1/python-3.11.1-amd64.exe"
  if [[ -f python-3.11.1-amd64.exe ]]; then
    wine python-3.11.1-amd64.exe
  else
    echo "Please download python-3.11.1-amd64.exe to this directory."
  fi

  #### Install NSIS 3.08
  wget "https://versaweb.dl.sourceforge.net/project/nsis/NSIS%203/3.08/nsis-3.08-setup.exe"
  if [[ -f nsis-3.08-setup.exe ]]; then
    wine nsis-3.08-setup.exe
  else
    echo "Please download nsis-3.08-setup.exe to this directory."
  fi

  #### Install scripts
  mkdir -p "$SCRIPTS"
  cp -rv "$SCRIPT_DIR/pve.bat" "$SCRIPTS/"
  ln -s "$SCRIPT_DIR/pywinc.sh" "$HOME/.local/bin/pywinc"
  rm -rf "$TMPDIR"
)

### Uninstall PyWinCLI
uninstall(){
  rm -rf "$PROFILE_DIR"  "$HOME/.local/bin/pywinc" "$SCRIPT_DIR"
}

if [[ -z $CMD ]]; then
  install
elif [[ $CMD == "-u" ]]; then
  uninstall
else
  $CMD
fi
